### PR TITLE
Implement Api

### PR DESCRIPTION
**Pull request by Taoshix**

### CODE REVIEW

File: include/Api.hpp [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+                auto hash = Hash((std::string)username, (std::string)password);
+                auto token = std::format("{}", hash);
+                authedUsers[userid] = token;
+                return crow::response(200, token);
+            }
+            else{
+                return crow::response(403, "Invalid Token");
+            }
+        });
+        CROW_ROUTE(app, "/register")
```
Comment (Mast3rwaf1z) [inline include/Api.hpp:86]: shouldn't auth and register be sorted under /user like: `/user/auth` and `/user/register`

File: include/Api.hpp [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+            }
+            auto token = z["token"].get<std::string>();
+            auto userid = UserIdFromToken(token);
+            if(authedUsers[userid] == token){
+                auto data = z.at("data");
+                for(auto i = data.begin(); i != data.end(); ++i){
+                    auto key = i.key();
+                    auto value = i.value().front().get<std::string>();
+                    db->userdata->modify(userid,{key},{value});
+                }
```
Comment (Mast3rwaf1z) [inline include/Api.hpp:124]:
```suggestion
                auto data = z["data"];
                for(auto keyJSON : data){
                    auto key = keyJSON.get<std::string>();
                    auto value = data[key];
                    db->userdata->modify(userid,{key},{value});
                }
```

File: include/Api.hpp [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+                for(auto i = data.begin(); i != data.end(); ++i){
+                    auto key = i.key();
+                    auto value = i.value().front().get<std::string>();
+                    for(auto setting : userSettings){
+                        auto settingsRow = db->settings->get(setting);
+                        if(settingsRow["key"] == key){
+                            db->settings->modify(setting,{"value"},{value});
+                        }
+                    }
+                }
```
Comment (Mast3rwaf1z) [inline include/Api.hpp:153]:
```suggestion
                auto data = z["settings"];
                auto userSettings = db->settings->get_where("userId",std::format("{}",userid));
                for(auto keyJSON : data){
                    auto key = keyJSON.get<std::string>();
                    auto value = data[key];
                    for(auto setting : userSettings){
                        auto settingsRow = db->settings->get(setting);
                        if(settingsRow["key"] == key){
                            db->settings->modify(setting,{"value"},{value});
                        }
                    }
                }
```

File: include/Api.hpp [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+            auto userid = UserIdFromToken(token);
+            if(authedUsers[userid] == token){
+
+                auto data = z.at("data").items();
+                auto journalid = db->journals->add({"comment","userId"}, {comment, std::format("{}",userid)});
+                for(auto i = data.begin(); i != data.end(); ++i){
+                    auto questionId = i.value().back().get<std::string>();
+                    auto answer = i.value().front().get<std::string>();
+                    db->answers->add({"answer","journalId","questionId"},{answer, std::format("{}",journalid), questionId});
+                }
```
Comment (Mast3rwaf1z) [inline include/Api.hpp:180]:
```suggestion
                auto data = z["data"];
                auto journalid = db->journals->add({"comment","userId"}, {comment, std::format("{}",userid)});
                for(auto JSONObj : data){
                    auto questionId = JSONObj["question"];
                    auto answer = JSONObj["answer"];
                    db->answers->add({"answer","journalId","questionId"},{answer, std::format("{}",journalid), questionId});
                }
```

File: include/Api.hpp [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+            crow::json::wvalue wv;
+            wv = std::move(vec);
+            return std::move(wv);
+        });
+        CROW_ROUTE(app, "/questions/defaults")
+        ([&]() {
+            auto questions = db->questions->get_where("tags","default");
+            std::vector<crow::json::wvalue> vec;
+            for(auto i = 0; i > questions.size(); i++){
+                auto question = db->questions->get(i);
```
Comment (Mast3rwaf1z) [inline include/Api.hpp:248]:
```suggestion
            for(auto question : questions){
```

File: include/Api.hpp [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+            crow::json::wvalue wv;
+            wv = std::move(vec);
+            return std::move(wv);
+        });
+        CROW_ROUTE(app, "/questions/get/<string>")
+        ([&](std::string tag) {
+            auto questions = db->questions->get_where("tags",tag);
+            std::vector<crow::json::wvalue> vec;
+            for(auto i = 0; i > questions.size(); i++){
+                auto question = db->questions->get(i);
```
Comment (Mast3rwaf1z) [inline include/Api.hpp:264]:
```suggestion
            for(auto question : questions){
```

File: tests/api-test.sh [hunk 7]
```diff
@@ -10,18 +10,18 @@ curl -X 'POST' -d "$authjson" $addr/register
...
 # /user/data/update
-curl -X 'POST' -d "{\"token\":\"$token\", \"data\":{}}" $addr/user/data/update
+curl -X 'POST' -d "{\"token\":\"$token\", \"data\":{\"agegroup\":\"42-69\", \"occupation\":\"school\"}}" $addr/user/data/update
 
 # /user/get/<uid>
 userdata=$(curl -X 'GET' $addr/user/get/$uid)
 
 # /journals/new
-curl -X 'POST' -d "{\"token\":\"$token\", \"data\":[{\"question\":\"q1\", \"answer\":\"a1\"}]}" $addr/journals/new
+curl -X 'POST' -d "{\"token\":\"$token\", \"comment\":\"Too many exams today.\", \"data\":[{\"question\":\"1\", \"answer\":\"a1\"},{\"question\":\"2\", \"answer\":\"a2\"}]}" $addr/journals/new
```
Comment (Mast3rwaf1z) [inline tests/api-test.sh:24]: maybe use a questionid fetched first?

**Review by Mast3rwaf1z — CHANGES_REQUESTED** [review]

File: include/Api.hpp [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+   public:
+    crow::SimpleApp app;
+    std::shared_ptr<Database> db;
+    std::map<int, std::string> authedUsers;
+    API(std::string path, int port) {
+        db = std::make_shared<Database>(path);
+
+        CROW_ROUTE(app, "/user/get/<int>")
+        ([&](int id) {
+            crow::json::wvalue x({});
```
[the remark this answers is not in the document]
Comment (Taoshix) [reply include/Api.hpp:19]: Reason for this is that CROW cannot parse arrays for some reason, but i still need crow to validate the parsed json, might reimplement and clean up later if we got time

File: include/Api.hpp (hunk 1 above)
Comment (Taoshix) [reply include/Api.hpp]: done

File: include/Api.hpp [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+            }
+            x = std::move(filteredUsers); //Funky solution to make it only return json array
+            return std::move(x);          //This is done due to "inconsistent type" errors appearing if not done
+        });
+        CROW_ROUTE(app, "/user/ids")
+        ([&]() {
+            crow::json::wvalue x;
+            auto users = db->users->get_where();
+            x = std::move(users);
+            return std::move(x);
```
Comment (Mast3rwaf1z) [inline include/Api.hpp:60]: nice

**Review by Mast3rwaf1z — APPROVED** [review]
;)

**Pull request merged**